File: backbone-go/domain/specifications/cursor.go
// Package specifications provides Criteria pattern for flexible queries
package specifications

import (
	"encoding/base64"
	"encoding/json"
	"fmt"
)

// cursorPayload is the JSON shape encoded inside a cursor token.
type cursorPayload struct {
	V  interface{} `json:"v"`  // sort field's value on the last row of the previous page
	ID string      `json:"id"` // that row's unique id
}

// EncodeCursor returns an opaque, URL-safe token identifying a position in
// a keyset-paginated result set — the alternative to LIMIT/OFFSET that
// doesn't degrade as the offset grows: instead of "skip 50,000 rows, then
// return 20", a keyset query says "give me the 20 rows right after this
// one", which Postgres can satisfy with a direct index seek regardless of
// how deep into the result set that is.
//
// sortValue is whatever the query's sort field held on the last row of the
// page just returned (anything encoding/json can round-trip: a number,
// string, bool, or time formatted as a string). id is that row's unique
// identifier — required because the sort field alone can tie (two products
// at the same price); without a tiebreaker, a tie could make keyset
// pagination skip or repeat rows across pages.
//
// Treat the token as a black box: its shape isn't part of the contract,
// and it is not signed or encrypted — a client can decode it, so don't put
// anything in sortValue that shouldn't be visible to them (it's already
// visible in the row that produced it, so this is normally a non-issue).
func EncodeCursor(sortValue interface{}, id string) (string, error) {
	raw, err := json.Marshal(cursorPayload{V: sortValue, ID: id})
	if err != nil {
		return "", fmt.Errorf("encode cursor: %w", err)
	}
	return base64.RawURLEncoding.EncodeToString(raw), nil
}

// DecodeCursor parses a token produced by EncodeCursor. sortValue comes
// back typed the way encoding/json decodes into interface{} (float64 for
// any JSON number, otherwise string/bool/nil) — the caller, which knows
// the sort field's real column type, is responsible for converting it to
// that type before binding it as a query parameter. That mirrors how
// FilterParam values are already converted from raw query-string text to
// a concrete type (parseScalar) before reaching a Specification: type
// coercion for a dynamic field belongs to the caller who knows the schema,
// not to this generic package.
func DecodeCursor(token string) (sortValue interface{}, id string, err error) {
	raw, err := base64.RawURLEncoding.DecodeString(token)
	if err != nil {
		return nil, "", fmt.Errorf("invalid cursor: %w", err)
	}
	var payload cursorPayload
	if err := json.Unmarshal(raw, &payload); err != nil {
		return nil, "", fmt.Errorf("invalid cursor: %w", err)
	}
	if payload.ID == "" {
		return nil, "", fmt.Errorf("invalid cursor: missing id")
	}
	return payload.V, payload.ID, nil
}

File: backbone-python/domain/specifications/cursor.py
"""
Cursor (keyset) pagination — opaque tokens identifying a position in a
result set.

The alternative to LIMIT/OFFSET that doesn't degrade as the offset grows:
instead of "skip 50,000 rows, then return 20", a keyset query says "give me
the 20 rows right after this one", which Postgres can satisfy with a direct
index seek regardless of how deep into the result set that is.

Same wire format as backbone-go's EncodeCursor/DecodeCursor (base64 of a
`{"v": ..., "id": ...}` JSON payload) — contracts match backbone Python and
Go exactly, same as everywhere else in this library.
"""
import base64
import json
from typing import Any, Tuple


def encode_cursor(sort_value: Any, id_: str) -> str:
    """Returns an opaque, URL-safe token identifying a position in a
    keyset-paginated result set.

    sort_value is whatever the query's sort field held on the last row of
    the page just returned; id_ is that row's unique identifier — required
    because the sort field alone can tie (two products at the same price);
    without a tiebreaker, a tie could make keyset pagination skip or repeat
    rows across pages.

    Treat the token as a black box: its shape isn't part of the contract,
    and it is not signed or encrypted — a client can decode it, so don't
    put anything in sort_value that shouldn't be visible to them (it's
    already visible in the row that produced it, so this is normally a
    non-issue).
    """
    raw = json.dumps({"v": sort_value, "id": id_}, default=str).encode("utf-8")
    return base64.urlsafe_b64encode(raw).rstrip(b"=").decode("ascii")


def decode_cursor(token: str) -> Tuple[Any, str]:
    """Parses a token produced by encode_cursor. Returns (sort_value, id).

    sort_value comes back typed the way json.loads decodes it (a number,
    string, bool, or None) — the caller, which knows the sort field's real
    column type, is responsible for converting it before binding it as a
    query parameter. That mirrors how FilterParser already converts raw
    query-string text to a concrete type before building a Specification:
    type coercion for a dynamic field belongs to the caller who knows the
    schema, not to this generic module.

    Raises ValueError if the token is malformed or missing an id.
    """
    try:
        padding = "=" * (-len(token) % 4)
        raw = base64.urlsafe_b64decode(token + padding)
        payload = json.loads(raw)
    except Exception as exc:
        raise ValueError(f"invalid cursor: {exc}") from exc

    id_ = payload.get("id") if isinstance(payload, dict) else None
    if not id_:
        raise ValueError("invalid cursor: missing id")
    return payload.get("v"), id_
